src/createSlice.ts

```diff
@@ -72,12 +77,36 @@ export interface CreateSliceOptions<
   extraReducers?: CaseReducers<S, any>
 }
 
-type CaseReducerActionPayloads<CR extends CaseReducers<any, any>> = {
+type PayloadActions<T extends keyof any = string> = Record<T, PayloadAction>
+
+type EnhancedCaseReducer<S, A extends PayloadAction> = {
+  reducer: CaseReducer<S, A>
+  prepare: PrepareAction<A['payload']>
+}
+
+type SliceCaseReducers<S, PA extends PayloadActions> = {
+  [T in keyof PA]: CaseReducer<S, PA[T]> | EnhancedCaseReducer<S, PA[T]>
+}
+
+type CaseReducerActions<CR extends SliceCaseReducers<any, any>> = {
   [T in keyof CR]: CR[T] extends (state: any) => any
-  ? void
-  : (CR[T] extends (state: any, action: PayloadAction<infer P>) => any
-    ? P
-    : void)
+    ? PayloadActionCreator<void>
+    : (CR[T] extends (state: any, action: PayloadAction<infer P>) => any
+        ? PayloadActionCreator<P>
+        : CR[T] extends { prepare: PrepareAction<infer P> }
+        ? PayloadActionCreator<P, string, CR[T]['prepare']>
```

Comment:
wat

Reply:
Oh, come on, THAT one isn't even so much different than before =D

Big difference is that it builds the complete ActionCreator type instead of just the payload, and also it now has 4 cases instead of 3.